### DISCUSSION TITLE
Narrowing an attribute type after validation

### DISCUSSION THREAD

**Discussion by gwerbin** [pydantic/pydantic#4823] [Question] [answered]
I have an attribute that can accept a `date` or `datetime` (or their string equivalents) as _input_, but should only ever produce a timezone-aware `datetime` as _output_.

My current implementation is:

```python
from datetime import date as Date, datetime as Datetime, timezone as Timezone

import pydantic


class TaskParams(pydantic.BaseModel):
    start_dt: Datetime | Date

    @pydantic.validator("start_dt")
    @classmethod
    def _ensure_tz_aware(cls, v: Datetime | Date) -> Datetime:
        if isinstance(v, Date):
            return Datetime(year=v.year, month=v.month, day=v.day, tzinfo=Timezone.utc)
        if v.tzinfo is None:
            return v.replace(tzinfo=Timezone.utc)
        return v
```

This design doesn't work well with static type checking, because the `TaskParams.start_dt` attribute is still annotated as `Datetime | Date` and not `Datetime`.

I added the `Date` in the union to instruct Pydantic to accept `datetime.date` objects, as well as strings of the form `'YYYY-MM-DD'`.

I could annotate the attribute with `Datetime` only and use a `pre=True` validator, but then the burden is on me to reimplement Pydantic's date and datetime-parsing logic, and possibly keep it up to date with future library updates. That seems like a poor design choice.

Another option is to define an accessor method or property that internally calls `assert isinstance(self.start_dt, Datetime)`, but that's also kind of an ugly workaround.

Ideally I would be able to instruct Pydantic somehow to parse/validate inputs as either `Datetime` or `Date`, leaving me free to set the attribute annotation as `start_dt: Datetime`. Is that possible somehow? I don't see an option for this in e.g. [`pydantic.Field`](https://docs.pydantic.dev/usage/schema/#field-customization).

Alternatively, is there an official recommended workaround for this scenario?

**Comment by gwerbin** [comment, marked as answer]
One workaround option is to use `if TYPE_CHECKING` as described in https://github.com/pydantic/pydantic/discussions/4725#discussioncomment-4337576:

```python
from datetime import date as Date, datetime as Datetime, timezone as Timezone
from typing import TYPE_CHECKING

import pydantic


class TaskParams(pydantic.BaseModel):
    if TYPE_CHECKING:
        start_dt: Datetime
    else:
        start_dt: Datetime | Date

    @pydantic.validator("start_dt")
    @classmethod
    def _ensure_tz_aware(cls, v: Datetime | Date) -> Datetime:
        if isinstance(v, Date):
            return Datetime(year=v.year, month=v.month, day=v.day, tzinfo=Timezone.utc)
        if v.tzinfo is None:
            return v.replace(tzinfo=Timezone.utc)
        return v

tp = TaskParams(start_dt="2022-01-01")  # Accepts a "date" (not "datetime") format
print(tp)  # Should be a datetime.datetime
if TYPE_CHECKING:
    reveal_type(tp.start_dt)  # Revealed type is datetime.datetime
```

**Reply by alecglen** [reply]
Hello from the future. This may work for the implementation but does not satisfy the static type checker. Do you know of a way to have it accept the Date/str input without falsely flagging it?
<img width="588" alt="image" src="https://github.com/user-attachments/assets/5736cdc6-2df5-4b9d-838b-9c17861a3b68" />
